File: server/src/main/java/com/uiptv/service/VodWatchStateService.java
package com.uiptv.service;

import com.uiptv.db.VodWatchStateDb;
import com.uiptv.model.Account;
import com.uiptv.model.Channel;
import com.uiptv.model.VodWatchState;

import java.util.List;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

import static com.uiptv.util.StringUtils.isBlank;

@SuppressWarnings("java:S6548")
public class VodWatchStateService {
    private final Set<VodWatchStateChangeListener> listeners = new CopyOnWriteArraySet<>();

    private VodWatchStateService() {
    }

    private static class SingletonHelper {
        private static final VodWatchStateService INSTANCE = new VodWatchStateService();
    }

    public static VodWatchStateService getInstance() {
        return SingletonHelper.INSTANCE;
    }

    public VodWatchState getVod(String accountId, String categoryId, String vodId) {
        if (isBlank(accountId) || isBlank(vodId)) {
            return null;
        }
        String canonicalVodId = canonicalizeVodId(vodId);
        VodWatchState exact = VodWatchStateDb.get().getByVod(accountId, normalize(categoryId), canonicalVodId);
        if (exact != null) {
            return exact;
        }
        VodWatchState latest = null;
        for (VodWatchState candidate : VodWatchStateDb.get().getByVod(accountId, canonicalVodId)) {
            if (candidate == null) {
                continue;
            }
            if (latest == null || candidate.getUpdatedAt() > latest.getUpdatedAt()) {
                latest = candidate;
            }
        }
        return latest;
    }

    public List<VodWatchState> getAllByAccount(String accountId) {
        if (isBlank(accountId)) {
            return List.of();
        }
        return VodWatchStateDb.get().getByAccount(accountId);
    }

    public boolean isSaved(String accountId, String categoryId, String vodId) {
        return getVod(accountId, categoryId, vodId) != null;
    }

    public void save(Account account, String categoryId, Channel channel) {
        if (account == null || isBlank(account.getDbId()) || channel == null || isBlank(channel.getChannelId())) {
            return;
        }
        VodWatchState state = new VodWatchState();
        state.setAccountId(account.getDbId());
        state.setCategoryId(normalize(categoryId));
        state.setVodId(canonicalizeVodId(channel.getChannelId()));
        state.setVodName(channel.getName());
        state.setVodCmd(channel.getCmd());
        state.setVodLogo(channel.getLogo());
        state.setUpdatedAt(System.currentTimeMillis());
        VodWatchStateDb.get().upsert(state);
        notifyListeners(account.getDbId(), state.getVodId());
    }

    public void remove(String accountId, String categoryId, String vodId) {
        if (isBlank(accountId) || isBlank(vodId)) {
            return;
        }
        String canonicalVodId = canonicalizeVodId(vodId);
        VodWatchStateDb.get().clear(accountId, normalize(categoryId), canonicalVodId);
        notifyListeners(accountId, canonicalVodId);
    }

    public void clearAll() {
        VodWatchStateDb.get().clearAll();
        notifyListeners("", "");
    }

    public void addChangeListener(VodWatchStateChangeListener listener) {
        if (listener != null) {
            listeners.add(listener);
        }
    }

    public void removeChangeListener(VodWatchStateChangeListener listener) {
        if (listener != null) {
            listeners.remove(listener);
        }
    }

    private void notifyListeners(String accountId, String vodId) {
        for (VodWatchStateChangeListener listener : listeners) {
            listener.onChanged(accountId, vodId);
        }
    }

    private String normalize(String value) {
        return value == null ? "" : value.trim();
    }

    private String canonicalizeVodId(String vodId) {
        String raw = vodId == null ? "" : vodId.trim();
        if (isBlank(raw) || !raw.contains(":")) {
            return raw;
        }
        String[] parts = raw.split(":");
        String last = "";
        for (int i = parts.length - 1; i >= 0; i--) {
            String p = normalize(parts[i]);
            if (!isBlank(p)) {
                last = p;
                break;
            }
        }
        return isBlank(last) ? raw : last;
    }
}

File: server/src/main/kotlin/com/uiptv/service/VodWatchStateService.kt
package com.uiptv.service

import com.uiptv.db.VodWatchStateDb
import com.uiptv.model.Account
import com.uiptv.model.Channel
import com.uiptv.model.VodWatchState
import com.uiptv.util.StringUtils
import java.util.concurrent.CopyOnWriteArraySet

object VodWatchStateService {
    private val listeners = CopyOnWriteArraySet<VodWatchStateChangeListener>()

    @JvmStatic
    fun getInstance(): VodWatchStateService = this
    fun getVod(accountId: String?, categoryId: String?, vodId: String?): VodWatchState? {
        if (StringUtils.isBlank(accountId) || StringUtils.isBlank(vodId)) {
            return null
        }
        val canonicalVodId = canonicalizeVodId(vodId)
        val normalizedAccountId = accountId.orEmpty()
        val exact = VodWatchStateDb.get().getByVod(normalizedAccountId, normalize(categoryId), canonicalVodId)
        if (exact != null) {
            return exact
        }
        var latest: VodWatchState? = null
        for (candidate in VodWatchStateDb.get().getByVod(normalizedAccountId, canonicalVodId)) {
            if (candidate != null && (latest == null || candidate.updatedAt > latest!!.updatedAt)) {
                latest = candidate
            }
        }
        return latest
    }
    fun getAllByAccount(accountId: String?): List<VodWatchState> {
        if (StringUtils.isBlank(accountId)) {
            return emptyList()
        }
        return VodWatchStateDb.get().getByAccount(accountId.orEmpty())
    }
    fun isSaved(accountId: String?, categoryId: String?, vodId: String?): Boolean = getVod(accountId, categoryId, vodId) != null
    fun save(account: Account?, categoryId: String?, channel: Channel?) {
        if (account == null || StringUtils.isBlank(account.dbId) || channel == null || StringUtils.isBlank(channel.channelId)) {
            return
        }
        val state = VodWatchState()
        state.accountId = account.dbId
        state.categoryId = normalize(categoryId)
        state.vodId = canonicalizeVodId(channel.channelId)
        state.vodName = channel.name
        state.vodCmd = channel.cmd
        state.vodLogo = channel.logo
        state.updatedAt = System.currentTimeMillis()
        VodWatchStateDb.get().upsert(state)
        notifyListeners(account.dbId.orEmpty(), state.vodId.orEmpty())
    }
    fun remove(accountId: String?, categoryId: String?, vodId: String?) {
        if (StringUtils.isBlank(accountId) || StringUtils.isBlank(vodId)) {
            return
        }
        val canonicalVodId = canonicalizeVodId(vodId)
        val normalizedAccountId = accountId.orEmpty()
        VodWatchStateDb.get().clear(normalizedAccountId, normalize(categoryId), canonicalVodId)
        notifyListeners(normalizedAccountId, canonicalVodId)
    }
    fun clearAll() {
        VodWatchStateDb.get().clearAll()
        notifyListeners("", "")
    }
    fun addChangeListener(listener: VodWatchStateChangeListener?) {
        if (listener != null) {
            listeners.add(listener)
        }
    }
    fun removeChangeListener(listener: VodWatchStateChangeListener?) {
        if (listener != null) {
            listeners.remove(listener)
        }
    }

    private fun notifyListeners(accountId: String, vodId: String) {
        listeners.forEach { it.onChanged(accountId, vodId) }
    }

    private fun normalize(value: String?): String = value?.trim().orEmpty()

    private fun canonicalizeVodId(vodId: String?): String {
        val raw = vodId?.trim().orEmpty()
        if (StringUtils.isBlank(raw) || !raw.contains(":")) {
            return raw
        }
        val last = raw.split(":").asReversed().firstOrNull { StringUtils.isNotBlank(it?.trim()) }?.trim().orEmpty()
        return if (StringUtils.isBlank(last)) raw else last
    }
}
